### ISSUE TITLE
[Bug]: File Descriptor Leak when upgrading from 2.35.0 to 2.36.0

### ISSUE THREAD

**Issue by NianticRyan** [apache/beam#24098]
### What happened?

Hi, 

I recently encountered an issue when upgrading my apache beam SDKs. I was attempting to upgrade from 2.30.0 to 2.40.0, and there was a file descriptor leak where approximately every time a new pipeline was started, java would create a new set of file descriptors to the libraries on the classpath. This resulted in many duplicate file descriptors being open and eventually hitting the maximum number of file descriptors.

I was able to track it down to when I bump from 2.35.0 to 2.36.0, and I am 99% certain that [this commit](https://github.com/apache/beam/commit/970bdc0ed7142f5263e9a78fc4d715d50539e7ef) is the source. 

The main difference in this code is that `classGraph.disableNestedJarScanning().addClassLoader(classLoader).scan(1).getClasspathFiles();` creates a new `Scanner` object with `performScan=true`, 
whereas `classGraph.disableNestedJarScanning().addClassLoader(classLoader).getClasspathFiles();` does the same with `performScan=false`. 
`PerformScan=true` runs an asynchronoous scan of the classpath, and `PerformScan=false` just generates a placeholder ScanResult with just the classpath. The scan itself is creating the fd leak.

The file leak itself is coming from the classgraph scanner. 
Here's one example of an open file descriptor
```
#697 /default/lib/guice-assistedinject-3.0.jar by thread:ClassGraph-worker-2 on Wed Nov 09 22:10:48 UTC 2022
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:244)
	at nonapi.io.github.classgraph.fileslice.FileSlice.<init>(FileSlice.java:134)
	at nonapi.io.github.classgraph.fileslice.FileSlice.<init>(FileSlice.java:178)
	at nonapi.io.github.classgraph.fastzipfilereader.PhysicalZipFile.<init>(PhysicalZipFile.java:87)
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$1.newInstance(NestedJarHandler.java:93)
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$1.newInstance(NestedJarHandler.java:90)
	at nonapi.io.github.classgraph.concurrency.SingletonMap.get(SingletonMap.java:189)
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$4.newInstance(NestedJarHandler.java:189)
	at nonapi.io.github.classgraph.fastzipfilereader.NestedJarHandler$4.newInstance(NestedJarHandler.java:154)
	at nonapi.io.github.classgraph.concurrency.SingletonMap.get(SingletonMap.java:189)
	at io.github.classgraph.ClasspathElementZip.open(ClasspathElementZip.java:162)
	at io.github.classgraph.Scanner$3.processWorkUnit(Scanner.java:595)
	at io.github.classgraph.Scanner$3.processWorkUnit(Scanner.java:567)
	at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkLoop(WorkQueue.java:246)
	at nonapi.io.github.classgraph.concurrency.WorkQueue.runWorkQueue(WorkQueue.java:161)
	at io.github.classgraph.Scanner.processWorkUnits(Scanner.java:342)
	at io.github.classgraph.Scanner.openClasspathElementsThenScan(Scanner.java:1047)
	at io.github.classgraph.Scanner.call(Scanner.java:1146)
	at io.github.classgraph.Scanner.call(Scanner.java:83)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

Our configuration is using kubernetes to kick off a google cloud dataflow using the apache beam sdk.
Please let me know if there is any other information I can provide.

Is anyone else hitting a similar issue?

### Issue Priority

Priority: 2

### Issue Component

Component: runner-core

**Comment by NianticRyan** [comment]
I did some more digging into the code. I believe this is a bug with the code in commit posted above.
The ScanReport needs to be closed before we return from the function or else the file descriptors are left open.

I cloned the repo and published my own version of the package with this code, and I'm not seeing the fd leaks anymore.

This wasn't an issue prior to 2.36 because as I noted in the original post, the code from 2.35 and lower does not do a full scan. It just generates a placeholder `ScanResult` with the classpath. A full scan actually goes in and opens every file and thus, the ScanReport needs to be closed after it is done - else we are left with unclosed fds. See this [link](https://github.com/classgraph/classgraph/blob/latest/src/main/java/io/github/classgraph/ClassGraph.java#L1563) in the docs of classgraph.
Will publish a PR with this fix soon.

```
  @Override
  public List<String> detect(ClassLoader classLoader) {
    ScanResult scanResult =  classGraph
            .disableNestedJarScanning()
            .addClassLoader(classLoader)
            .scan(1);
    List<File> classpathContents = scanResult.getClasspathFiles();
    scanResult.close();

    return classpathContents.stream().map(File::getAbsolutePath).collect(Collectors.toList());
  }
```

**Comment by Abacn** [comment]
I see the original bug https://github.com/classgraph/classgraph/issues/614 is resolved. Bump version to >4.8.145 and roll back the workaround of #16275 may resolve the issue

**Comment by NianticRyan** [comment]
I'm not convinced that will solve the issue. That github post is complaining about non-deterministic behavior between scan() and scan(1). Rolling back the workaround removes the scan entirely.  I can't say exactly why the full scan is needed, but I can theorize that in the original commit [#16275](https://github.com/apache/beam/pull/16275) the committer meant to say that the full scan is needed, but they could not use scan() and rather needed to use scan(1) due to the bug in classgraph.

scan() still runs a full scan of the classpath and objects. It just does it with multiple threads rather than scan(1) which only does it with 1 thread. If we change to using that, then we still have a bug because we need to close the ScanReport which comes back from the scan() call.

**Comment by kennknowles** [comment]
@steveniemitz the author of the commit might have ideas

**Comment by kennknowles** [comment]
`ScanResult` extends `Autocloseable` so the fix would be something like this:

```java
@Override
  public List<String> detect(ClassLoader classLoader) {
    List<File> classpathContents;
    try (ScanResult scanResult =  classGraph
            .disableNestedJarScanning()
            .addClassLoader(classLoader)
            .scan(1)) {
      classpathContents = scanResult.getClasspathFiles();
    }
    scanResult.close();

    return classpathContents.stream().map(File::getAbsolutePath).collect(Collectors.toList());
  }
```

@NianticRyan or @steveniemitz would you like to open a PR?

**Comment by NianticRyan** [comment]
I have a bit more context on the issue now. 

Prior to commit #16275, we were just calling `getClasspathFiles()` on the Classgraph instance which generates the `ScanResult` internally within the ClassGraph library. Since they are generating the report internally, they are closing it. 
However, now we are calling `scan(1)` which returns a `ScanResult`. Since we are now holding the handle to `ScanResult`, we need to close it.

So yeah, either rolling back to the previous change and upgrading to the latest ClassGraph or making a change such as I or @kennknowles suggested.

If possible, I'd like someone else to submit the PR. I've been trying to submit a PR over the past couple of weeks and github's automated systems have suspended my account twice so far for reasons that they won't tell me. Every time they suspend my account, it also hides this bug report.

**Comment by steveniemitz** [comment]
😲

I can open a PR for this, I'll do what @kennknowles suggested for now I think.

**Comment by NianticRyan** [comment]
Excellent. Thanks so much for the help. Any idea when the next release will be? My team is blocked on doing the upgrade until this is fixed, so it'd be nice to have a timeline of when we'll be able to upgrade.

**Comment by steveniemitz** [comment]
You can override the detector to another (custom) implementation if you want via PipelineOptions, I forget which option it's on but I know we used to override it to work around this before my original PR.

**Comment by NianticRyan** [comment]
Oh that is a great tip. I'll go ahead and use that as a workaround to enable the upgrade until the next release.

**Comment by NianticRyan** [comment]
One other interesting datapoint here...
So I implemented the workaround suggested and am passing in a custom ClassPathScanningResourcesDetector in the options. This has resolved the issue where we see file descriptor leaks whenever a new pipeline is started. 

However, I am now seeing a new leak where every 2 hours, I see a jump in the file descriptors used. It's always around the same amount (which is around how many jar files we have on our classpath). Also interestingly, it happens on xx:42 no matter what time I start my instances.

Going to try upgrading classgraph and reverting back to the original implementation to see if I still observe this issue.

Here's a list of file descriptor reference open times on my instance.

```
   1023 Dec  6 19:43
    280 Dec  6 19:53
    370 Dec  6 19:56
      1 Dec  6 20:01
      1 Dec  6 20:07
      1 Dec  6 20:49
    780 Dec  6 21:42
      1 Dec  6 21:54
      1 Dec  6 21:59
    778 Dec  6 23:42
      1 Dec  6 23:43
      1 Dec  6 23:47
      1 Dec  7 00:17
      1 Dec  7 00:18
    777 Dec  7 01:42
    778 Dec  7 03:42
      1 Dec  7 03:43
      1 Dec  7 03:48
      1 Dec  7 03:58
      1 Dec  7 04:03
      1 Dec  7 04:04
    780 Dec  7 05:42
    779 Dec  7 07:42
      1 Dec  7 07:50
      1 Dec  7 07:53
      1 Dec  7 07:54
    780 Dec  7 09:42
    779 Dec  7 11:42
      1 Dec  7 11:43
      1 Dec  7 11:44
      1 Dec  7 11:45
      1 Dec  7 11:48
      1 Dec  7 11:50
    780 Dec  7 13:42
    768 Dec  7 15:42
      1 Dec  7 16:34
      1 Dec  7 16:40
    777 Dec  7 16:42
      1 Dec  7 16:43
      1 Dec  7 17:28
      1 Dec  7 18:32
      1 Dec  7 18:33
      1 Dec  7 18:34
    774 Dec  7 18:42
      1 Dec  7 19:00
      ```

**Comment by steveniemitz** [comment]
Are the jars being uploaded during that time?  There was a "bug" I fixed with HTTP connection thrashing while uploading artifacts (https://github.com/apache/beam/pull/22432), thats in 2.42.

**Comment by NianticRyan** [comment]
Ah nevermind. It turns out that was another regular running dataflow that I did not apply the fixed options to.

**Comment by kennknowles** [comment]
So is this resolved? Could the issue be closed?

**Comment by NianticRyan** [comment]
No, we still need a pull request to fix the bug. I've just been able to work around this locally using the suggestion @steveniemitz suggested.
